File: backend/waf/middleware/circuit_breaker.py
import time

from fastapi import HTTPException


class CircuitBreaker:
    def __init__(self, failure_threshold: int = 5, timeout: float = 60.0):
        self.failure_threshold = failure_threshold
        self.timeout = timeout
        self.failure_count = 0
        self.last_failure_time: float | None = None
        self.state = "CLOSED"

    async def call(self, func, *args, **kwargs):
        if self.state == "OPEN":
            if self.last_failure_time and (time.time() - self.last_failure_time) > self.timeout:
                self.state = "HALF_OPEN"
            else:
                raise HTTPException(status_code=503, detail="Circuit breaker OPEN - upstream unavailable")

        try:
            result = await func(*args, **kwargs)
            self.on_success()
            return result
        except Exception as e:
            self.on_failure()
            raise e

    def on_success(self):
        self.failure_count = 0
        self.state = "CLOSED"

    def on_failure(self):
        self.failure_count += 1
        self.last_failure_time = time.time()
        if self.failure_count >= self.failure_threshold:
            self.state = "OPEN"


circuit_breaker = CircuitBreaker()

File: proxy/internal/middleware/circuit_breaker.go
package middleware

import (
	"sync"
	"sync/atomic"
	"time"

	"github.com/rs/zerolog/log"
)

type state int32

const (
	stateClosed   state = 0
	stateHalfOpen state = 1
	stateOpen     state = 2
)

type CircuitBreaker struct {
	failureThreshold int
	resetTimeout     time.Duration
	state            int32
	failureCount     int32
	lastFailureTime  atomic.Value
	mu               sync.Mutex
}

func NewCircuitBreaker(threshold int, resetTimeout time.Duration) *CircuitBreaker {
	cb := &CircuitBreaker{
		failureThreshold: threshold,
		resetTimeout:     resetTimeout,
	}
	cb.lastFailureTime.Store(time.Time{})
	return cb
}

func (cb *CircuitBreaker) Process(ctx *Context) {
	if atomic.LoadInt32(&cb.state) == int32(stateOpen) {
		lastFail := cb.lastFailureTime.Load().(time.Time)
		if time.Since(lastFail) > cb.resetTimeout {
			atomic.StoreInt32(&cb.state, int32(stateHalfOpen))
			log.Debug().Msg("circuit breaker half-open")
		} else {
			ctx.Blocked = true
			ctx.ThreatType = "circuit_breaker"
			httpError(ctx.Response, "Service temporarily unavailable")
			return
		}
	}
}

func (cb *CircuitBreaker) RecordFailure() {
	count := atomic.AddInt32(&cb.failureCount, 1)
	cb.lastFailureTime.Store(time.Now())

	if count >= int32(cb.failureThreshold) {
		atomic.StoreInt32(&cb.state, int32(stateOpen))
		log.Warn().Int32("failures", count).Msg("circuit breaker opened")
	}
}

func (cb *CircuitBreaker) RecordSuccess() {
	atomic.StoreInt32(&cb.failureCount, 0)
	atomic.StoreInt32(&cb.state, int32(stateClosed))
}
